### ISSUE TITLE
Support more streamlined authentication experience for returning users

### ISSUE THREAD

**Issue by yi-gu** [fedidcg/FedCM#429]
We have heard from both RP and IDPs that they would like the authentication experience to be more streamlined for users who **have already created a federated account** with them. While this is achievable today by RPs setting long-lived session cookies, such a solution may be discouraged due to security reasons such as cookie theft. Managing long-lived cookies is challenging for many RPs and the cost gets proportionally increased with longer cookie lifetime.

**Comment by yi-gu** [comment]
We believe that the browser can help to reduce user friction at the re-authentication moment for RPs/IDPs without paying such security cost. In addition, since the new streamlined re-authn experience only applies to returning users who have granted permission for the RP-IDP communication in the past, we believe that this can be done without reducing privacy.

A quick overview of how the proposed  “auto re-authn” works:
1. Prerequisite: a user has already created an account on rp.example with idp.example in the browser, and has not cleared site data for either site
1. When the user returns to rp.example, they may no longer have an active session with the RP.  e.g. 1P session cookie expired
1. If “auto re-authn” is available (see details below), instead of showing the explicit sign-in UI with the “Continue as” call-to-action button, the browser can automatically fetch the token from idp.example and return it to rp.example
1. If successful, the user will be automatically re-authenticated into the RP. The browser can display some UI in the “auto re-authn” flow to give users some context of this feature.

## Proposals
Auto re-authn is for “returning users” only. In addition, it’s gated by several other conditions. We define **“Auto re-authn is available”** if all the following are true

1. An RP has opted in to auto re-authn AND
1. Only one account is eligible for auto re-authn AND
1. Auto re-authn is not in cooldown state

### Auto re-authn flow overview
<p align="center">
     <img width="467" src="https://github.com/fedidcg/FedCM/blob/main/proposals/images/auto-re-authn-flow.png?raw=true"/>
</p>

### API proposal
An RP can specify that they’d like to use auto re-authn by adding the autoReauthn field to the existing API:
```
const cred = await navigator.credentials.get({
    identity: {
        providers: [{
            configURL: "https://idp.example/fedcm.json",
            clientId: "1234",
        }],
        autoReauthn: true, // default to false
    },
});
```
When it’s true, and all other conditions are met, the browser can resolve the promise with the id token without explicit user action.


## Privacy Consideration
The major difference between the existing authentication flow and the auto re-authn flow is that the latter does not require explicit user permission for authentication. However, an explicit user permission to allow the RP-IDP communication MUST have been granted in the browser in the past (e.g. the browser stores such permission locally after a previous FedCM grant). i.e. the new feature only works “post permission”. 

From privacy’s perspective, this “post permission” feature does not introduce new privacy risks because an explicit user permission to allow RP-IDP communication has been granted in the browser in the past. Thus, bad actors can track user visits on an RP already. e.g. during the last time the user logged in with FedCM, RP and IDP could have already exchanged an identifier such that the RP could just make a fetch to the IDP and reveal relevant information to IDP anyway.

## Security Consideration
Technically speaking, the concept “auto re-authn” is achievable on the RP land. e.g. when a user is signing in to an RP with an IDP, the RP can set a long-lived cookie (e.g. 400 days) such that the user is always signed in to the RP as long as they revisit the RP within 400 days. However, while being able to automatically sign returning users back in is a promising feature for RPs, having to extend 1P cookies may be suboptimal or even impractical for some RPs. The proposed API on the other hand:
1. provides all RPs the opportunity of “auto re-authn” without jeopardizing their cookie policy
 - Having long-lived session cookies are hard because an RP has to deal with cookie theft etc. which gets proportionally harder the longer it lives
 - It’s cheaper to operate by delegating it to the IDP (auto re–authn works when the user is signed in to the IDP, i.e. IDP needs to manage long-lived session cookies but RPs don’t)
2. does not introduce any security risk similar to privacy since it’s “post permission”

## Considered alternatives
### mediation: silent
The Credential Management API [supports]([url](https://www.w3.org/TR/credential-management-1/#dom-credentialmediationrequirement-silent)) `mediation: silent` to achieve similar goals. In particular, the intended usage of `mediation: silent` is to
> support "Keep me signed-into this site" scenarios, where a developer may wish to silently obtain credentials if a user should be automatically signed in, but to delay bothering the user with a sign-in prompt until they actively choose to sign-in.

While the concepts are similar, there are some notable differences from our use case:
1. Fallback strategy. With `mediation: silent`, the operation will return null rather than involving the user when the requested credential is unavailable. e.g. when a user visits an RP for the first time on a new device, if `mediation: silent` is used, the user won't see any UI as if the API is never called. For FedCM, we'd like to fallback to the explicit authentication flow if the user is not "returning" such that RPs don't need to choose different mediation modes. It aligns with our activation strategy: pulling as much effort as possible from RP side to the browser side.

1. Stored credential. `mediation: silent` works when there's a single local credential that gets stored previously. However, FedCM does not store any Credential object (e.g. IdentityCredential) locally and always fetches from IDP on the fly. 

1. sign-in v.s. authentication. In a successful FedCM flow, we resolve the promise with an id token as part of the authentication. However, an RP can require extra steps before actually signing the user in. e.g. 2-step verification. i.e. FedCM API cannot guarantee that the user is signed in post authentication which deviates a bit from the intended purpose of `mediation: silent`. 

Reusing the existing `mediation: silent` without monkey patching seems suboptimal from both API and ergonomics's perspective. It may be feasible to introduce another mediation mode for our use case. That said, the differences above are mostly FedCM specific. We should probably also add `IdentityCredential` to the Credential Management API if this option is desirable. 

## FedCM extensions consideration
### Multiple IDP support
FedCM is adding multiple IDP support (explainer) which may have an intersection with auto re-authn. e.g. if both IDP1 and IDP2 are eligible for FedCM, how does FedCM handle auto re-authn?

Similar to the “single IDP with multiple accounts” use case, we can choose to trigger auto re-authn only if the user has one returning account across the IDPs. e.g.

|                              | Account with IDP 1 is new | Account with IDP 1 is returning |
| ------------- |-------------| -----|
| **Account with IDP 2 is new**       | No auto re-authn          | Auto re-authn with IDP 1        |
| **Account with IDP 2 is returning** | Auto re-authn with IDP 2  | No auto re-authn                |

## UX Consideration
Note: this section is UA specific and subject to change.

### UX proposal
As mentioned in the motivation section, auto re-authn provides streamlined user experience for users who have already created an account on the website without paying privacy or security cost. In the meantime we believe that providing some informative UI to users during the auto re-authn flow could give users more context.

When a user comes back to a website after some time, there are three possible scenarios:
1. the user is still signed in to the RP (e.g. the 1P RP cookie has not expired). RP/IDP should not call FedCM in this case so this is out of scope of FedCM
1. if auto re-authn is unavailable, the user will see the typical UI asking for their permission to proceed. e.g. the user will be presented with a UI with the “Continue as” button.
1. if auto re-authn is available, the browser will automatically send the token request to the IDP and show the UI with animation as if the user has clicked the “Continue as” button. e.g.
<p align="center">
     <img width="467" src="https://github.com/fedidcg/FedCM/blob/main/proposals/images/auto-re-authn-ui.png?raw=true"/>
</p>

#### Auto re-authn cooldown
When a user is automatically signed in, they may sign out if they want to switch an account or prefer to anonymously visit the site this time. To avoid the site from calling FedCM with auto re-authn immediately after the sign-out (which could lead to a dead loop), we disable auto re-authn within 10 mins from the last authentication.

#### Implicit user opt-out
When auto re-authn is triggered, there’s a UI with animation being displayed for 3 seconds. While the UI is not CTA, there’s currently a “X” button on it similar to the “Verifying…” UI in the non auto re-authn flow. If the “X” button is clicked, we may consider it as a signal of the user trying to end the flow and opt-out the user implicitly. However, according to the data we gathered so far, many users close the “Verifying…” UI today after they explicitly grant permission by clicking the “Continue as” button. i.e. user closing the UI doesn’t necessarily mean that they want to end the flow, it could also mean that they want to dismiss the UI sooner to continue browsing the website.

Therefore, we propose to monitor the “X” button close rate and only implicitly opt out users if auto re-authn has a statistically impact on the close rate. 

### Explicit user sign-out
We propose not to tie auto re-authn with explicit user sign-out. e.g. after a user signs out, we don’t disable auto re-authn for that RP. 

A user may sign out from an RP due to several reasons:
1. they have finished their business on the RP
1. they are using a shared/public computer
1. they want to browse the RP anonymously
1. they were signed in with a wrong account

For case #1, there’s no obvious signal that the user doesn’t want to sign in to the RP next time
For case #2, if the user either has cleared the cookies/browsing history, or signed out from the IDP to switch accounts, auto re-authn won’t be triggered next time.
For cases #3 and #4, the “auto re-authn cooldown” feature is able to honor their sign-out intention and won’t cause a dead loop.

From security ux’s perspective, since this feature is “post permission”, even though there’s a chance that a user may be automatically signed in unintentionally, there’s no privacy/security leak. In addition, the user is able to update their auto re-authn preferences from settings.
#### Considered alternative
While disabling auto re-authn after explicit user sign-out of the RP is also an acceptable UX, there’s no reliable way for the browser to detect such user action. Technically speaking the browser can expose a sign-out API such that the RP can “notify” the browser that the user has signed out and the browser can disable auto re-authn accordingly. That being said, this sign-out API is not a reliable way either since RPs are less incentivized to use it. Such inconsistency among RPs could lead to bad UX as well.
### Cancellable auto re-authn
When auto re-authn flow is triggered, it’s possible that a user wants to cancel it. One possible solution is that before the browser fetches the token automatically, it shows a UI for users to cancel the auto re-authn flow. If users don’t take any action in 5s, the browser proceeds with the token fetch. While this solution provides more controls to some users, it does raise accessibility concerns due to the call-to-action UI being timed out. We could use a longer timeout value but that would negate the benefits of auto re-authn.

**Comment by yi-gu** [comment]
@bvandersloot-mozilla 
Hi Ben,
We are thinking about running an origin trial of this API and will send out an "Intent to Experiment" to blink-dev. Since this is not **shipping**, we are wondering whether a PR is needed to get you feedback as discussed the other day. Note that the API shape may change before shipping based on the experiment.

**Comment by bvandersloot-mozilla** [comment]
@martinthomson : thoughts on if feedback is needed before an Intent to Experiment?

**Comment by bvandersloot-mozilla** [comment]
To the proposal: I think I prefer using mediation modes- this feels like re-inventing that wheel.

To your concerns in the considered alternatives:
1. This is exactly the purpose of `mediation: optional`.
2. I think it would be good to add this! This was part of my `federated3` proposal I was going to spin out into an atomic PR once my large editorial patches were done.
3. I'm not sure I follow, but I think the core of the argument is "FedCM API cannot guarantee that the user is signed in post authentication which deviates a bit from the intended purpose of mediation: silent." I disagree with this. The intended purpose of the CM API is to get a credential, not guarantee that the user is signed in. I don't see how the semantics here are any different than existing uses of FedCM (or any other Credential). The RP can always require other steps to log in.

Also, 
> We should probably also add IdentityCredential to the Credential Management API if this option is desirable.
This has already happened :) https://w3c.github.io/webappsec-credential-management/#sctn-cred-type-registry

**Comment by yi-gu** [comment]
Thanks Ben for the quick response!

> 1. This is exactly the purpose of `mediation: optional`.

Agreed. On the other hand, `mediation: optional` is the default mode for `get` which means all RPs are opt-in to auto re-authn by default and they have to opt out explicitly with `mediation: required` if they don't want this feature.

For one, it changes the default behavior for FedCM sign-in flow. We need to think about what it means to future extensions like multiple IDP support and other potential use cases we have discussed about at FedIDCG.
In addition, this seems like a backwards incompatible change (from UX's perspective and how developers recognize FedCM in general) so we need to evaluate whether it's reasonable to proceed (unlike IdpSigninStatus API which I have solid reason to make that backwards incompatible change).

> 3. I'm not sure I follow, but I think the core of the argument is "FedCM API cannot guarantee that the user is signed in post authentication which deviates a bit from the intended purpose of mediation: silent." I disagree with this. The intended purpose of the CM API is to get a credential, not guarantee that the user is signed in. I don't see how the semantics here are any different than existing uses of FedCM (or any other Credential). The RP can always require other steps to log in.

That's a fair point. I suppose we can ask for some wording changes in CM API to make it more generic if needed.

> Also,
> 
> > We should probably also add IdentityCredential to the Credential Management API if this option is desirable.
> > This has already happened :) https://w3c.github.io/webappsec-credential-management/#sctn-cred-type-registry

TIL :)

As mentioned earlier, origin trial is a way for us to understand whether the current proposal (API, UX etc.) makes sense such that we can course correct before shipping. While we are totally open with the "mediation mode" path and will be actively thinking it through, we'd like to proceed with the "autoReauthn" path for this round of origin trial to get some feedback on other fronts. Does it make sense?

Regardless the API shape, can I assume that there's no "objection" from you for the feature itself? Of course we need to agree on the API before shipping but that could be a separate conversation when we get there. WDYT?

**Comment by martinthomson** [comment]
@bvandersloot-mozilla:

> thoughts on if feedback is needed before an Intent to Experiment?

I think that we'd like to get a little further into this before we can be happy.

I'd say that if this were more clearly the right design, then it might be easier to say "no problem", but I don't see that being the case here.

Let's look at this from a privacy perspective.  Once the RP has obtained a user-bound identifier from the IdP, then the two sites are able to link activity from that person.  That means that re-login flows can indeed be done without re-prompting, assuming that cookies have not been cleared on either site[^1].

It's not clear to me whether this is just about recovering from lost RP cookies, or whether it is really about complying with IdP policies about token validity and periodic re-authorization (of the RP by the IdP).

The proposed Boolean parameter doesn't really cover cases where there are potentially multiple identities involved though.  I would prefer that the RP indicate directly which identity they want to re-authenticate, including the associated IdP.  This would allow for multiple linked identities (potentially across different IdPs) without additional friction on this flow.  There are ways to make this state more robust than a login cookie, which might expire for good reasons.

If the goal is to offer an option to choose a different identity, the browser is in a position to manage that.  It will have better information.  Though I will note that we might want to avoid offering new identities in general once one has been established; we might prefer to only show the chooser in the case where the original identity is no longer available for use.

There are lots of things in this proposal that I don't think are necessary (rate limiting re-authorization and showing browser UX being the ones that spring to mind).

[^1]: On this point, I would like to see the spec be clearer about what cookie clearing on either site does to login state, but let's set that aside for now.

**Comment by yi-gu** [comment]
Thanks Martin for the feedback! Some comments inline:

> That means that re-login flows can indeed be done without re-prompting, assuming that cookies have not been cleared on either site[1](#user-content-fn-1-7bff082092e4a194b8de2da60038fdc0).

Great to hear that we are aligned on the privacy side! Yes if a user has cleared cookies/site data on either RP or IDP, the previous permission should be revoked. I listed it as the `Prerequisite` of the proposal but will make it clear going forward.

> 
> It's not clear to me whether this is just about recovering from lost RP cookies, or whether it is really about complying with IdP policies about token validity and periodic re-authorization (of the RP by the IdP).

The mail goal of this API is not about solving the RP cookie problem (if any). Rather, it's to reduce unnecessary user friction "post permission". As mentioned in the "Problem" section, any explicit confirmation after user returning to an RP does not provide privacy/security benefit like it does during the sign-up moment.

> 
> The proposed Boolean parameter doesn't really cover cases where there are potentially multiple identities involved though. I would prefer that the RP indicate directly which identity they want to re-authenticate, including the associated IdP. This would allow for multiple linked identities (potentially across different IdPs) without additional friction on this flow. There are ways to make this state more robust than a login cookie, which might expire for good reasons.
> 

That makes sense. Rps should have the opportunity to specify which identity (or account) they'd like to re-authenticate. We have proposed another API (l[ogin_hint](https://github.com/fedidcg/FedCM/issues/426)) which I believe can help with this situation. i.e. for Rps who want to re-authenticate a specific account, they could add `login_hint` to the `get` call. For those who are OK with any account, they could skip the `login_hint` bit and the browser could select the only returning account based on its own knowledge. Actually, with `login_hint`, we could potentially support multiple returning accounts if only one of them is "hinted" by the RP. WDYT?

> If the goal is to offer an option to choose a different identity, the browser is in a position to manage that. It will have better information. Though I will note that we might want to avoid offering new identities in general once one has been established; we might prefer to only show the chooser in the case where the original identity is no longer available for use.

Understood! There are a couple of ways to reduce the chance for duplicated accounts. Skipping the chooser when there's one that's available is definitely a compelling solution. Another possible path is to show which account is new or returning on the chooser such that the user could be better informed about the situation. Will report back. That said, do you think this could be UA specific UX that's orthogonal to the re-authn API?

> 
> There are lots of things in this proposal that I don't think are necessary (rate limiting re-authorization and showing browser UX being the ones that spring to mind).
> 

For "showing browser UX", IIUC, do you prefer skipping browser UI during re-authentication?
For "rate limiting re-authorization", could you elaborate a bit? Did you mean re-authentication?

**Comment by martinthomson** [comment]
> with login_hint, we could potentially support multiple returning accounts if only one of them is "hinted" by the RP. WDYT?

This is a little ambiguous to me.  The form of login_hint is such that you can't always rely on it being an identity.  That is, in OIDC, login_hint is poorly defined to the extent that it could just be a message that the RP sends to the IdP.  That is clearly unacceptable in this context, so we might need to tighten this up more and define it as an identity.

At that point, the notion of automatic login is something the browser can look after.  The Boolean parameter is not needed.  If the identity is correct, then you are in one of two states: 

1. the site has not been authorized, so you need to talk to the user before acquiring a token; you might promote that identifier somehow, but that is down to browser choice
2. the site was previously authorized, so you should avoid annoying the user and just get a token

If the identity is not correct, then you proceed with the full flow.  Note that if the user has previously authorized a different identifier, this might be a way to force the browser into getting a different identifier.  I'm not sure if this is something I like or not.  I would prefer if RPs had only had one identity from one IdP, but I'd need to hear more about use cases to be confident in that position.

> do you think this could be UA specific UX that's orthogonal to the re-authn API?

Yeah, we might offer guidance, but, as above, my preference is to avoid prompting where possible.  If a prompt is shown, we might nudge people toward a previous choice rather than expand RP-IdP communication options, but mostly this is to avoid inadvertent account explosions.

> For "showing browser UX", IIUC, do you prefer skipping browser UI during re-authentication?

I refer to your little popup that has no user interaction, other than maybe a cancellation.  The mock shows this as fully inside the content area, with suggests that sites can render their own UX if they like.  Also, the idea that a user might cancel a login is a bit pessimistic about IdP performance.  Yes, IdPs can be quite slow, but they don't have to be.  The cancellation only works well if they are very slow, but even then it has poor accessibility properties.  I'd say you'd need excellent reflexes and fine motor control to hit that small button in time.

>  For "rate limiting re-authorization", could you elaborate a bit? Did you mean re-authentication?

By this, I refer to your proposed 10 minute timer.  Let the IdP manage that.  Note that I said re-authorization because the authentication happens at the IdP; the RP is authorized by the IdP (or by the user).

**Comment by yi-gu** [comment]
> This is a little ambiguous to me. The form of login_hint is such that you can't always rely on it being an identity. That is, in OIDC, login_hint is poorly defined to the extent that it could just be a message that the RP sends to the IdP. That is clearly unacceptable in this context, so we might need to tighten this up more and define it as an identity.

Sorry for not being clear. In case you haven't seen our `login_hint` proposal, the hint we are defining is slightly different from OIDC. i.e. it currently only accepts "account id" or "email" which are clear identifiers. We are also considering using other fields in [IdentityProviderAccount](https://fedidcg.github.io/FedCM/#dictdef-identityprovideraccount). Please see the "Extensibility" section in issue #426. If we tight the hint to reliable identifiers, does it mitigate your concern? It's possible that we should consider a rename for `login_hint` given the potential confusion.

> 1. the site has not been authorized, so you need to talk to the user before acquiring a token; you might promote that identifier somehow, but that is down to browser choice

Agreed.

> 2. the site was previously authorized, so you should avoid annoying the user and just get a token

As discussed earlier, this is definitely an option for Chrome that we want to iterate on.
> 
> If the identity is not correct, then you proceed with the full flow. Note that if the user has previously authorized a different identifier, this might be a way to force the browser into getting a different identifier.

Agree that this is a bit tricky. Basically it means that the browser has observed a single returing account that it can auto re-authenticate with but the RP has hinted a different account. We haven't really thought about this use case through but my intuition is that we should respect the RP's choice and NOT proceed with auto re-authn. We are also proposing `isRequired` for `login_hint` that allows the RP to specify whether they accept other accounts that are different from the one hinted (there are use cases that only specific account is accepted by the RP, e.g. accounts in enterprise and edu). Perhaps we could trigger different flows based on the `isRequired` bit?

> the idea that a user might cancel a login is a bit pessimistic about IdP performance.

As you mentioned (also listed in the proposal) we won't proceed with a cancelable auto re-authn. (just to be clear, the thinking was about only sending the token request if a user doesn't cancel in N seconds so it should be orthogonal to IdP performance? Anyway it sounds like that this UX should be off the table)

> By this, I refer to your proposed 10 minute timer. Let the IdP manage that.

We are introducing the cooldown to make sure that a user can logout and switch accounts if needed. Totally understand that you prefer "one identity per RP" and we share the same principle such that we'd suggest users to reuse their existing account other than creating new ones. That said, according to some user study we ran, there are users who have multiple identities on an RP for a variety of reasons so they may need to switch accounts every now and then.

Ideally, once a user is signed out of the RP explicitly (e.g. via a "Sign me out" button), the RP should not call FedCM with auto re-authn such that the user can switch accounts. However that's not guaranteed and the user will be signed in with the wrong account again when the RP fails to call FedCM properly. While it's indeed a programming error on the RP/IDP side, maybe the browser can take a position in this case to help with that? e.g. disabling auto re-authn for some time so they can switch accounts.

Again, for the origin trial, we'd like to gather feedback from all parties and evaluate the decisions we made before shipping. WDYT?

**Comment by bvandersloot-mozilla** [comment]
I've made a draft of what I was talking about RE the mediation parameter in #439. It is a draft because I don't think it is ready, however it is worth having something concrete to argue against.

**Comment by Bojhan** [comment]
How would this work in the use case where the user has two IDP's (my university and my research institute). They try both with the publishers site, and find out that only one works. 

How do we make it clear to the user, either one has to be removed (but only for this publishers site) or the user can choose which to tie initial re-auth with?

**Comment by yi-gu** [comment]
Before the user has completed a FedCM flow with either account, we always show both accounts without auto re-authn.

If the user has created an account with one of them (either university or research institute), next time the user is visiting the same publishers site, the publisher site can choose to trigger auto re-authn with that selected account. 

if the user has created two accounts with both of them on the publisher site, auto re-authn will not be triggered and instead the user will need to manually select which one they want to proceed *this time*.

Note that the decisions are made based on the browser's knowledge of whether the user has created an account or not. i.e. if the user clears cookies or browsing history etc., the browser *can* choose to reset that knowledge.
